File: Python/16-AdvancedTopics/4-KMPAlgorithm(PatternMatching)/KMPAlgorithm(PatternMatching).py
def compute_lps_array(pattern):
    """
    Computes the Longest Prefix Suffix (LPS) array for the given pattern.

    The LPS array is used in the KMP algorithm to efficiently find occurrences of a
    pattern within a text.  lps[i] stores the length of the longest prefix of
    pattern[0...i] which is also a suffix of pattern[0...i].

    For example:
    - For pattern "ABABDABACD", the LPS array is [0, 0, 1, 2, 0, 1, 2, 3, 0, 0].
    - For pattern "AAAA", the LPS array is [0, 1, 2, 3].
    - For pattern "ABCDE", the LPS array is [0, 0, 0, 0, 0].

    Args:
        pattern: The pattern string (str).

    Returns:
        A list (list of int) representing the LPS array.
        Returns an empty list if the pattern is empty.
    """
    m = len(pattern)
    lps = [0] * m  # Initialize LPS array with 0s.  lps[i] will store the longest prefix
    # that is also a suffix of pattern[0...i].
    length = 0  # Length of the previous longest prefix suffix. This variable keeps
    # track of the length of the longest prefix that is also a suffix
    # of the current prefix being considered.
    i = 1  # Index for the current character in the pattern.  We start from index 1
    # because lps[0] is always 0.

    # Base case: lps[0] is always 0 because there is no proper prefix for the first character.
    lps[0] = 0

    # Calculate LPS array
    while i < m:
        if pattern[i] == pattern[length]:
            # If the current character matches the character at the 'length' position,
            # it means we have found a longer prefix suffix.
            length += 1  # Increment the length of the longest prefix suffix.
            lps[i] = length  # Store the length in the LPS array.
            i += 1  # Move to the next character in the pattern.
        else:
            # If the characters do not match, we need to check for shorter prefix suffixes.
            if length != 0:
                # This is the crucial part of the KMP algorithm.  Instead of starting
                # the comparison from the beginning of the pattern, we use the
                # information stored in the LPS array to backtrack efficiently.
                # We set 'length' to the value stored in lps[length-1]. This tells us
                # the length of the next smaller prefix that is also a suffix.
                length = lps[length - 1]
                # Note: We do not increment i here. We stay at the same character i
                # and compare it with the shorter prefix.
            else:
                # If length is 0, it means there is no prefix that is also a suffix,
                # so we set lps[i] to 0 and move to the next character.
                lps[i] = 0
                i += 1
    return lps


def kmp_search(text, pattern):
    """
    Implements the Knuth-Morris-Pratt (KMP) algorithm to find occurrences of a
    pattern within a text.

    Args:
        text: The text string to search within (str).
        pattern: The pattern string to search for (str).

    Returns:
        A list (list of int) containing the starting indices of all occurrences
        of the pattern in the text. Returns an empty list if no occurrences are found.
    """
    n = len(text)
    m = len(pattern)
    occurrences = []  # List to store starting indices of matches

    if m == 0:
        return [0] if n == 0 else [] # handle empty pattern edge case. If pattern is empty
        # and text is empty, return [0]. If pattern is empty
        # and text is not empty, return [].
    if n == 0:
        return [] # handle empty text edge case. If text is empty, there can be no occurrences.

    # Compute the LPS array for the pattern.  This is a pre-processing step that
    # optimizes the search.
    lps = compute_lps_array(pattern)

    i = 0  # Index for the text.
    j = 0  # Index for the pattern.

    while i < n:
        if text[i] == pattern[j]:
            # If the current characters match, move to the next characters in both
            # the text and the pattern.
            i += 1
            j += 1

        if j == m:
            # If j reaches the length of the pattern, it means we have found a complete match.
            occurrences.append(i - j)  # Store the starting index of the match.
            # i - j gives the index in 'text' where the match starts.
            j = lps[j - 1]  # Reset j to check for overlapping occurrences.  For example,
            # if text is "AAAAAAA" and pattern is "AAAA", after finding the
            # first match at index 0, we reset j to lps[3] which is 3,
            # so we find the next match at index 1.
        elif i < n and text[i] != pattern[j]:
            # If the characters do not match and we haven't reached the end of the text:
            if j != 0:
                # This is where the LPS array is used for optimization.  Instead of
                # resetting j to 0 and starting the comparison from the beginning of
                # the pattern, we use the LPS array to find the longest prefix of the
                # pattern that is also a suffix of the already matched part of the text.
                j = lps[j - 1]
            else:
                # If j is 0, it means there was no previous match, so we move to the
                # next character in the text.
                i += 1
    return occurrences

if __name__ == "__main__":
    # Example usage of the kmp_search function
    text1 = "ABABDABACDABABCABAB"
    pattern1 = "ABABCABAB"
    result1 = kmp_search(text1, pattern1)
    print(f"Text: {text1}")
    print(f"Pattern: {pattern1}")
    print(f"Occurrences: {result1}")  # Output: Occurrences: [10]

    text2 = "AAAAAAA"
    pattern2 = "AAAA"
    result2 = kmp_search(text2, pattern2)
    print(f"\nText: {text2}")
    print(f"Pattern: {pattern2}")
    print(f"Occurrences: {result2}")  # Output: Occurrences: [0, 1, 2, 3]

    text3 = "THIS IS A TEST TEXT"
    pattern3 = "TEST"
    result3 = kmp_search(text3, pattern3)
    print(f"\nText: {text3}")
    print(f"Pattern: {pattern3}")
    print(f"Occurrences: {result3}") # Output: Occurrences: [10]

    text4 = "aabaabaaa"
    pattern4 = "aaba"
    result4 = kmp_search(text4, pattern4)
    print(f"\nText: {text4}")
    print(f"Pattern: {pattern4}")
    print(f"Occurrences: {result4}") # Output:  [0, 3]

File: JavaScript/16-AdvancedTopics/4-KMPAlgorithm(PatternMatching)/KMPAlgorithm(PatternMatching).js
/**
 * KMP Algorithm (Pattern Matching) - 5 Approaches in JavaScript
 *
 * The Knuth-Morris-Pratt (KMP) algorithm is an efficient string-searching algorithm
 * that searches for occurrences of a "word" (pattern) within a main "text string"
 * by employing the observation that when a mismatch occurs, the word itself embodies
 * sufficient information to determine the next position where matching could begin,
 * thus bypassing re-examination of previously matched characters.
 */

// Helper function to compute the Longest Proper Prefix Suffix (LPS) array
// This is the core of the KMP algorithm.  The LPS array helps to avoid
// redundant comparisons during the search.
function computeLPSArray(pattern) {
    const m = pattern.length;
    const lps = new Array(m).fill(0); // Initialize LPS array with 0s
    let len = 0; // Length of the previous longest prefix suffix
    let i = 1; // Index for the pattern string

    lps[0] = 0; // LPS[0] is always 0

    while (i < m) {
        if (pattern[i] === pattern[len]) {
            // If characters match, increment len and assign it to LPS[i]
            len++;
            lps[i] = len;
            i++;
        } else {
            // If characters do not match
            if (len !== 0) {
                // Important: This is the key optimization.  Instead of going back to the beginning
                // of the pattern, we use the information in the LPS array to jump to the
                // correct position.
                len = lps[len - 1]; // Also can be written as len = lps[--len];
                // Note that we do not increment i here.
            } else {
                // If len is 0, there is no proper prefix which is also a suffix,
                // so set LPS[i] to 0 and increment i.
                lps[i] = 0;
                i++;
            }
        }
    }
    return lps;
}

/**
 * Approach 1: Classic KMP Algorithm
 * - Computes the LPS array.
 * - Uses the LPS array to efficiently search for the pattern in the text.
 */
function kmpSearchClassic(text, pattern) {
    const n = text.length;
    const m = pattern.length;

    if (m === 0) {
        return []; // Handle empty pattern case
    }

    const lps = computeLPSArray(pattern);
    let i = 0; // Index for text
    let j = 0; // Index for pattern
    const occurrences = []; // Array to store starting indices of matches

    while (i < n) {
        if (pattern[j] === text[i]) {
            // If characters match, increment both indices
            j++;
            i++;
        }

        if (j === m) {
            // If the entire pattern is matched
            occurrences.push(i - j); // Store the starting index of the match
            j = lps[j - 1]; // Prepare to check for the next occurrence
        } else if (i < n && pattern[j] !== text[i]) {
            // If characters do not match
            if (j !== 0) {
                // Use LPS array to find the next position to compare
                j = lps[j - 1];
            } else {
                // If j is 0, no prefix matches, so move to the next character in text
                i++;
            }
        }
    }
    return occurrences;
}

/**
 * Approach 2: KMP Search with Detailed Logging
 * - Same as the classic KMP, but with added console logs for illustration.
 */
function kmpSearchWithLogging(text, pattern) {
    const n = text.length;
    const m = pattern.length;

    if (m === 0) {
        console.log("Pattern is empty. Returning empty array.");
        return [];
    }

    const lps = computeLPSArray(pattern);
    console.log("LPS Array:", lps);

    let i = 0;
    let j = 0;
    const occurrences = [];

    console.log("Text:", text);
    console.log("Pattern:", pattern);

    while (i < n) {
        console.log(`Comparing text[${i}] ('${text[i]}') and pattern[${j}] ('${pattern[j]}')`);
        if (pattern[j] === text[i]) {
            j++;
            i++;
            console.log("Match! j =", j, "i =", i);
        }

        if (j === m) {
            console.log("Pattern found at index:", i - j);
            occurrences.push(i - j);
            j = lps[j - 1];
            console.log("j updated to:", j, "from LPS array");
        } else if (i < n && pattern[j] !== text[i]) {
            if (j !== 0) {
                console.log("Mismatch! j =", j, "i =", i);
                j = lps[j - 1];
                console.log("j updated to:", j, "from LPS array");
            } else {
                i++;
                console.log("Mismatch and j is 0. Incrementing i. i =", i);
            }
        }
    }
    console.log("Occurrences:", occurrences);
    return occurrences;
}

/**
 * Approach 3: KMP Search with Early Exit on First Match
 * - Stops searching after finding the first occurrence of the pattern.
 */
function kmpSearchFirstOccurrence(text, pattern) {
    const n = text.length;
    const m = pattern.length;

    if (m === 0) {
        return -1; // Return -1 to indicate no match (consistent with not found)
    }

    const lps = computeLPSArray(pattern);
    let i = 0;
    let j = 0;

    while (i < n) {
        if (pattern[j] === text[i]) {
            j++;
            i++;
        }

        if (j === m) {
            return i - j; // Return the starting index of the first match
        } else if (i < n && pattern[j] !== text[i]) {
            if (j !== 0) {
                j = lps[j - 1];
            } else {
                i++;
            }
        }
    }
    return -1; // Return -1 if no match is found
}

/**
 * Approach 4: KMP Search using a Class
 * - Encapsulates the KMP algorithm within a class for better organization.
 */
class KMP {
    constructor(pattern) {
        this.pattern = pattern;
        this.m = pattern.length;
        this.lps = this.m > 0 ? computeLPSArray(pattern) : []; // Compute LPS only if pattern is not empty.
    }

    search(text) {
        const n = text.length;
        if (this.m === 0) return [];

        let i = 0;
        let j = 0;
        const occurrences = [];

        while (i < n) {
            if (this.pattern[j] === text[i]) {
                j++;
                i++;
            }

            if (j === this.m) {
                occurrences.push(i - j);
                j = this.lps[j - 1];
            } else if (i < n && this.pattern[j] !== text[i]) {
                if (j !== 0) {
                    j = this.lps[j - 1];
                } else {
                    i++;
                }
            }
        }
        return occurrences;
    }
}

/**
 * Approach 5: KMP Search with Overlapping Matches Count
 * - Returns the number of overlapping matches.
 */
function kmpSearchOverlappingCount(text, pattern) {
    const n = text.length;
    const m = pattern.length;

    if (m === 0) {
        return 0;
    }

    const lps = computeLPSArray(pattern);
    let i = 0;
    let j = 0;
    let count = 0;

    while (i < n) {
        if (pattern[j] === text[i]) {
            j++;
            i++;
        }

        if (j === m) {
            count++;
            j = lps[j - 1]; // Allow overlapping matches.
        } else if (i < n && pattern[j] !== text[i]) {
            if (j !== 0) {
                j = lps[j - 1];
            } else {
                i++;
            }
        }
    }
    return count;
}

// Example Usage and Output
const text = "ABABDABACDABABCABAB";
const pattern = "ABABCABAB";

console.log("Text:", text);
console.log("Pattern:", pattern);

console.log("\nApproach 1: Classic KMP Search");
console.log("Occurrences:", kmpSearchClassic(text, pattern)); // Output: [ 10 ]

console.log("\nApproach 2: KMP Search with Detailed Logging");
kmpSearchWithLogging(text, pattern); // Output: (Detailed logs showing the process)

console.log("\nApproach 3: KMP Search - First Occurrence");
console.log("First Occurrence at index:", kmpSearchFirstOccurrence(text, pattern)); // Output: 10

const kmpInstance = new KMP(pattern);
console.log("\nApproach 4: KMP Search using a Class");
console.log("Occurrences:", kmpInstance.search(text)); // Output: [ 10 ]

console.log("\nApproach 5: KMP Search - Overlapping Matches Count");
console.log("Overlapping Matches Count:", kmpSearchOverlappingCount(text, pattern)); // Output: 1
